src/batch.jl

```diff
@@ -0,0 +1,199 @@
+"""
+    BatchIntegrand(f!, y::AbstractVector, x::AbstractVector, max_batch=typemax(Int))
+
+Constructor for a `BatchIntegrand` accepting an integrand of the form `f!(y,x) = y .= f.(x)`
+that can evaluate the integrand at multiple quadrature nodes using, for example, threads,
+the GPU, or distributed-memory. The `max_batch` keyword roughly limits the number of nodes
+passed to the integrand, though at least `4*order+2` nodes will be used by the GK rule.
+The buffers `y,x` must both be `resize!`-able since the number of evaluation points may vary
+between calls to `f!`.
+"""
+struct BatchIntegrand{F,Y,X}
+    # in-place function f!(y, x) that takes an array of x values and outputs an array of results in-place
+    f!::F
+    y::Y
+    x::X
+    max_batch::Int # maximum number of x to supply in parallel
+    function BatchIntegrand(f!, y::AbstractVector, x::AbstractVector, max_batch::Integer=typemax(Int))
+        max_batch > 0 || throw(ArgumentError("max_batch must be positive"))
+        return new{typeof(f!),typeof(y),typeof(x)}(f!, y, x, max_batch)
+    end
+end
+
+"""
+    BatchIntegrand(f!, y, x; max_batch=typemax(Int))
+
+Constructor for a `BatchIntegrand` with pre-allocated buffers.
+"""
+BatchIntegrand(f!, y, x; max_batch::Integer=typemax(Int)) =
+    BatchIntegrand(f!, y, x, max_batch)
+
+"""
+    BatchIntegrand(f!, y::Type, x::Type=Nothing; max_batch=typemax(Int))
+
+Constructor for a `BatchIntegrand` whose range type is known. The domain type is optional.
+Array buffers for those types are allocated internally.
+"""
+BatchIntegrand(f!, Y::Type, X::Type=Nothing; max_batch::Integer=typemax(Int)) =
```

Comment:
```suggestion
BatchIntegrand(f!, ::Type{Y}, ::Type{X}=Nothing; max_batch::Integer=typemax(Int)) where {X,Y} =
```
You need to make these explicit parameters [so that Julia specializes](https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing).  Shouldn't affect anything else.